### PR TITLE
add support for reading only parts of uvfits files

### LINKED ISSUE #304: Add option to only read header & antenna table data from uvfits
Avoid reading in the data for time/memory load in cases where it's not really needed.

### PR DESCRIPTION
**Pull request by bhazelton**
This addresses the uvfits part of issue #255
This resolves #304

### CODE REVIEW

**Comment by bhazelton** [comment]
I tweaked the tutorial a bit as part of this PR, you can see how it looks here: http://pyuvdata.readthedocs.io/en/uvfits_memmap/tutorial.html

**Comment by bhazelton** [comment]
ok, @adampbeardsley I think this is ready for review.

File: pyuvdata/uvfits.py [hunk 1]
```diff
@@ -104,113 +87,272 @@ def read_uvfits(self, filename, run_check=True, check_extra=True,
...
+                    raw_data_array = vis_hdu.data.data[blt_inds, 0, 0, :, :, :]
+                    raw_data_array = raw_data_array[:, np.newaxis, :, :, :]
+                if freq_frac < 1:
+                    raw_data_array = raw_data_array[:, :, freq_inds, :, :]
+                if pol_frac < 1:
+                    raw_data_array = raw_data_array[:, :, :, pol_inds, :]
+            elif freq_frac == min_frac:
+                if vis_hdu.header['NAXIS'] == 7:
+                    raw_data_array = vis_hdu.data.data[:, 0, 0, :, freq_inds, :, :]
+                    # for some reason the axes get transposed. fix the ordering
```
Comment (plaplant) [inline pyuvdata/uvfits.py]: Why were the axes transposed? Is it because we're adding these axes that weren't on disk?

Comment (bhazelton) [reply pyuvdata/uvfits.py]: I really don't understand why they were transposed. But it seemed to be consistently transposed, so I corrected for it.

File: pyuvdata/uvfits.py [hunk 2]
```diff
@@ -279,24 +421,107 @@ def read_uvfits(self, filename, run_check=True, check_extra=True,
...
+        """
+
+        hdu_list = fits.open(filename, memmap=True)
+        vis_hdu = hdu_list[0]  # assumes the visibilities are in the primary hdu
+
+        self._get_parameter_data(vis_hdu)
+
+        del(vis_hdu)
+
+    def read_uvfits_data(self, filename, antenna_nums=None, antenna_names=None,
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvfits.py]: That's a good point. I need to think a little about it, but putting in some checks is a good idea.

File: pyuvdata/uvfits.py [hunk 3]
```diff
@@ -279,24 +421,107 @@ def read_uvfits(self, filename, run_check=True, check_extra=True,
...
+            # don't read in the data. This means the object is incomplete,
+            # but that may not matter for many purposes.
+            return
+
+        # Now read in the data
+        self._get_data(vis_hdu, antenna_nums, antenna_names, ant_str,
+                       ant_pairs_nums, frequencies, freq_chans, times, polarizations,
+                       blt_inds, run_check, check_extra, run_check_acceptability)
+
+    def read_uvfits_metadata(self, filename):
```
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvfits.py]: Yes, I think it should raise a warning or error if the data array exists.

File: pyuvdata/uvfits.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (bhazelton) [reply pyuvdata/uvfits.py]: I added an option to read the metadata even if it already exists and defaulted it to True, so you can bypass it but you have to do it on purpose.

**Comment by bhazelton** [comment]
@plaplant I made some changes to try to address your comments. Can you take another look when you get a chance?

File: pyuvdata/uvfits.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (plaplant) [reply pyuvdata/uvfits.py]: Yep, that sounds good to me.

File: pyuvdata/uvfits.py (hunk 1 above)
Comment (plaplant) [reply pyuvdata/uvfits.py]: Okay. I'm a little uncomfortable not knowing the base reason for the transpose, but I suppose as long as the unit tests pass so we can catch any changes in behavior down the road, I'm okay with it.

Comment (bhazelton) [reply pyuvdata/uvfits.py]: I'm somewhat uncomfortable about this too. I'm not sure how to figure it out. The one thing that made me feel better is that the check method that gets run by default after the read checks that the array shapes are correct. Of course that's not sufficient if different axes have the same length. Let me do a little more googling...

**Comment by bhazelton** [comment]
@plaplant I figured out the transposition issue (I was selecting along too many axes at once). There are no more mysterious transpositions in the code. Can you take one more quick look?

**Pull request merged**